draft/celery-5-high-level-architecture.rst

```diff
@@ -0,0 +1,1216 @@
+
+.. code-block:: pycon
+
+  >>> from uuid import UUID
+  >>> from celery import task, event
+  >>> from myapp.models import User, AccountManager
+  >>> @task
+  ... def send_welcome_email(user_id, email):
+  ...   send_email(email=email, contents="hello, welcome", subject="welcome") # Send a welcome email to the user...
+  ...   User.objects.filter(pk=user_id).update(welcome_email_sent=True)
+  >>> @task
+  ... def notify_account_manager(user_id, email):
+  ...   account_manager = AccountManager.objects.assign_account_manager(user_id)
+  ...   send_email(email=account_manager.email, contents="you have a new user to attend to", subject="Alert") # Send an email to the account manager...
+  >>> @event
+  ... class UserRegistered:
+  ...   user_id: UUID
+  ...   email: str
+  >>> UserRegistered.subscribe(send_welcome_email)
+  >>> UserRegistered.subscribe(notify_account_manager)
+  >>> UserRegistered.delay(user_id=1, email='foo@bar.com')  # Calls both send_welcome_email and notify_account_manager with the provided arguments.
+
+These architectural building blocks will aid us in creating a better messaging
+system. To encourage `ubiquitous language`_, we will be using them in this
+document and in Celery 5's codebase as well.
+
+Observability
+-------------
+
+One of Celery 5's goals is to be :term:`observable <Observability>`.
+
+Each Celery component will record statistics, provide trace points for
+application monitoring tools and distributed tracing tools and emit log messages
+when appropriate.
+
+Metrics
++++++++
+
+Celery stores and publishes metrics which allows our users to debug their
+applications more easily and spot problems.
+
+By default each worker will publish the metrics to a dedicated queue.
+
+Other methods such as publishing them to StatsD is also possible using the
+provided extension point.
+
+Trace Points
+++++++++++++
+
+Celery provides trace points for application monitoring tools and distributed
+tracing tools.
+
+This allows our users to spot and debug performance issues.
+
+Logging
++++++++
+
+All log messages must be structured.
+:term:`Structured logs <Structured Logging>` provide context for our users
+which allows them to debug problems more easily and aids the developers
+to resolve bugs in Celery.
+
+The structure of a log message is determined whenever a component
+is initialized.
+
+During initialization, an attempt will be made to detect how the component
+lifecycle is managed.
+If all attempts are unsuccessful, the logs will be formatted using
+:term:`JSON` and will be printed to stdout.
+
+Celery will provide an extension point for detection of different
+runtimes.
+
+.. admonition:: Example
+
+  If a component's lifecycle is managed by a SystemD service,
+  Celery will detect that the `JOURNAL_STREAM`_ environment variable
+  is set when the process starts and use it's value to transmit structured
+  data into `journald`_.
+
+Whenever Celery fails to log a message for any reason it publishes a command
```

Comment:
If I understand correctly what is described here, I am not sure this is a good approach. A failure on a logging facility, could escalate to flooding a broker with undelivered log entries.

Reply:
Loosing observability on the other hand is also a problem.

We could devise a strategy which buffers logs which have failed to be written and only if it failed N number of times we publish a message to the broker.